src/atlas/interpolation/method/binning/Binning.cc

```diff
@@ -5,193 +5,262 @@
+    for (auto bufPtr = buffer.data(); bufPtr != buffer.data() + buffer.size(); bufPtr += sizeof(Triplet)) {
+        auto triplet = Triplet{};
+        std::memcpy(&triplet, bufPtr, sizeof(Triplet));
+        functor(triplet);
+    }
 }
 
+}  // namespace
 
-void Binning::do_setup(const Grid& source,
-                       const Grid& target,
-                       const Cache&) {
-  ATLAS_NOTIMPLEMENTED;
-}
-
-void Binning::do_setup(const FunctionSpace& source,
-                       const FunctionSpace& target,
-                       const Cache&) {
-  ATLAS_NOTIMPLEMENTED;
+Binning::Binning(const Config& config): Method(config) {
+    const auto* conf = dynamic_cast<const eckit::LocalConfiguration*>(&config);
+    ATLAS_ASSERT(conf, "config must be derived from eckit::LocalConfiguration");
+    interpAncillaryScheme_ = conf->getSubConfiguration("scheme");
+    adjoint_               = conf->getBool("adjoint", false);
 }
 
 
-void Binning::do_setup(const FunctionSpace& source,
-                       const FunctionSpace& target) {
-  ATLAS_TRACE("atlas::interpolation::method::Binning::do_setup()");
-
-  using Scalar  = eckit::linalg::Scalar;
-  using Index   = eckit::linalg::Index;
-  using Triplet = eckit::linalg::Triplet;
-  using SMatrix = eckit::linalg::SparseMatrix;
+void Binning::do_setup(const Grid& source, const Grid& target, const Cache&) {
+    ATLAS_NOTIMPLEMENTED;
+}
 
-  source_ = source;
-  target_ = target;
+void Binning::do_setup(const FunctionSpace& source, const FunctionSpace& target, const Cache&) {
+    ATLAS_NOTIMPLEMENTED;
+}
 
-  if (target_.size() == 0) {
-    return;
-  }
 
-  // note that the 'source' grid for the low-to-high regridding (interpolation)
-  // is the 'target' grid for high-to-low regridding (binning) and
-  // the 'target' grid for the low-to-high regridding (interpolation) is the
-  // 'source' grid for the for high-to-low regridding (binning)
-  const auto& fs_source_interp = target_;
-  const auto& fs_target_interp = source_;
+void Binning::do_setup(const FunctionSpace& source, const FunctionSpace& target) {
+    ATLAS_TRACE("atlas::interpolation::method::Binning::do_setup()");
 
-  const auto interp = Interpolation(
-    interpAncillaryScheme_, fs_source_interp, fs_target_interp);
-  auto smx_interp_cache = interpolation::MatrixCache(interp);
+    source_ = source;
+    target_ = target;
 
-  auto smx_interp = smx_interp_cache.matrix();
+    if (target_.size() == 0) {
+        return;
+    }
 
-  auto eckit_smx_interp = make_non_owning_eckit_sparse_matrix(smx_interp);
-  SMatrix smx_interp_tr(eckit_smx_interp); // copy
-  smx_interp_tr.transpose(); // transpose the copy in-place
+    const auto interp              = Interpolation(interpAncillaryScheme_, target_, source_);
+    const auto interpMatrixStorage = interpolation::MatrixCache(interp).matrix();
+    auto interpMatrix              = linalg::make_non_owning_eckit_sparse_matrix(interpMatrixStorage);
```

Comment:
You can make this more general by only using View and Storage:
```
SparseMatrixView<Scalar,Index> interpMatrixView = atlas::linalg::make_host_view<Scalar,Index>(interpMatrixStorage);
```
and refactor the transpose+halo-exchange routine to create a new MatrixStorage from the view:
```c++
SparseMatrixStorage Binning::transposeAndHaloExchange(SparseMatrixView<Scalar,Index> interpMatrix) const;
```
In this function you use an iterator which is not implemented as of yet in `SparseMatrixView`, so a regular (also more efficient) for loop would be required then, as here: https://github.com/ecmwf/atlas/blob/develop/src/atlas/linalg/sparse/SparseMatrixToTriplets.h#L24